File: Week-02/Theory_Assignment_01/1.py
from collections import defaultdict

def is_bidirectional(graph):
    for node,neighbours in graph.items():
        for neighbour in neighbours:
            if node not in graph[neighbour]:
                return False
    return True

# n = number of nodes
# m = number of edges
n,m = map(int,input().split())

# Create an adjacency list representation of the graph:
graph = defaultdict(list)

for i in range(m):
    u,v = map(int,input().split())
    graph[u].append(v)


#check if the graph is bidirectional:
if is_bidirectional(graph):
    print("The graph is bidirectional.")
else:
    print("This graph is not bidirectional.")


""" 
#for not bidirectional input:
5 6
0 1
1 2
1 3
2 0
3 4
4 1

#for directional input:
4 8
0 1
1 0
1 2
2 1
1 3
3 1
2 3
3 2
"""

File: Week-02/Theory_Assignment_01/1.cpp
#include <bits/stdc++.h>

/*
0---->1------>3
^    /^      /
|  /   \    /
| V     \  v
2         4

5 6

0 1
1 2
2 0
1 3
3 4
4 1
-------------
Bidirectional graph:
4 8

0 1
1 0
1 2
2 1
1 3
3 1
2 3
3 2


*/
using namespace std;
int main()
{
    int n, m;
    cin >> n >> m;
    vector<vector<int>> adj_list(n);
    for (int i = 0; i < m; i++)
    {
        int u, v;
        cin >> u >> v;
        adj_list[u].push_back(v);
    }

    bool is_bidirectional = true;
    for (int u = 0; u < n; u++)
    {
        for (int v : adj_list[u])
        {
            bool reverse_edge_found = false;
            for (int w : adj_list[v])
            {
                if (w == u)
                {
                    reverse_edge_found = true;
                    break;
                }
            }

            if (!reverse_edge_found)
            {
                is_bidirectional = false;
                break;
            }
        }
        if (!is_bidirectional)
        {
            break;
        }
    }

    if (is_bidirectional)
    {
        cout << "The graph is bi-directional\n";
    }
    else
    {
        cout << "The graph is not bi-directional\n";
    }

    return 0;
}